lib/cli/src/commands/run/wasi.rs

```diff
@@ -225,21 +248,164 @@ impl Wasi {
         Ok((wasi_env, instance))
     }
 
+    // Runs the Wasi process
+    pub fn run(run: RunProperties, store: Store) -> Result<i32> {
+        let tasks = run.ctx.data(&store).tasks().clone();
+
+        // The return value is passed synchronously and will block until the result is returned
+        // this is because the main thread can go into a deep sleep and exit the dedicated thread
+        let (tx, rx) = std::sync::mpsc::channel();
+
+        // We run it in a blocking thread as the WASM function may otherwise hold
+        // up the IO operations
+        tasks.task_dedicated(Box::new(move || {
+            Self::run_with_deep_sleep(run, store, tx, None);
+        }))?;
+        rx.recv()
+            .expect("main thread terminated without a result, this normally means a panic occurred within the main thread")
+    }
+
+    // Runs the Wasi process (asynchronously)
+    pub fn run_with_deep_sleep(
+        run: RunProperties,
+        mut store: Store,
+        tx: Sender<Result<i32>>,
+        rewind_state: Option<(RewindState, Result<(), Errno>)>,
```

Comment:
It feels weird that we are passing a result as an argument.
I think we can pass an `Option<Errno>` instead?

Reply:
That's subjective as an Option does not normally represent an error but in this case it does. I'm not hung on it though so happy to change it if you like but its more a personal preference I think

Reply:
agreed - I'll leave it like this for now